src/osd/OSDMapRecovery.cc

```diff
@@ -0,0 +1,463 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+/*
+ * Ceph - scalable distributed file system
+ *
+ * Copyright (C) 2016 SUSE LINUX GmbH
+ *
+ * This is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License version 2.1, as published by the Free Software
+ * Foundation.  See file COPYING.
+ *
+ */
+
+#include <iostream>
+#include <errno.h>
+
+#include "osd/OSD.h"
+#include "osd/OSDMap.h"
+#include "osd/OSDMapRecovery.h"
+
+#include "os/ObjectStore.h"
+
+#include "msg/Messenger.h"
+#include "msg/Message.h"
+#include "mon/MonClient.h"
+
+#include "messages/MOSDMap.h"
+#include "messages/MMonGetOSDMap.h"
+
+#include "common/ceph_context.h"
+#include "common/errno.h"
+#include "include/assert.h"
+#include "common/config.h"
+
+#define dout_subsys ceph_subsys_osd
+#undef dout_prefix
+#define dout_prefix _prefix(_dout, this)
+
+static ostream& _prefix(std::ostream* _dout, OSDMapRecovery *osdmr) {
+  return *_dout << "osd." << osdmr->osd->get_nodeid()
+                << " osdmr(" << osdmr->get_state() << ") ";
+}
+
+/* OSDMapRecovery attempts to recover broken osd maps by asking them from
+ * the monitors. Recovering osd maps may mask other corruptions. This should
+ * be considered solely as a last resort.
+ *
+ * This is how it goes:
+ *
+ * 'init' gets basic requirements set up
+ *
+ *  - a dedicated messenger, so we don't mess with the OSDs own client ms
+ *  - a mon client, dully authenticated
+ *
+ */
+int OSDMapRecovery::init()
+{
+  int auth_attempts = 0;
```

Comment:
I could, but then we'd be crossing the initialization with the gotos a few lines down.